src/grep.js

```diff
@@ -62,23 +86,101 @@ function _grep(options, regex, files) {
       }
     } else {
       var lines = contents.split('\n');
+      var matches = [];
+
+      //  If before or after not given but context is, update values
+      if (options.context) {
+        if (options.beforeContext === false) {
+          options.beforeContext = options.context;
+        }
+        if (options.afterContext === false) {
+          options.afterContext = options.context;
+        }
+      }
+
       lines.forEach(function (line, index) {
         var matched = line.match(regex);
         if ((options.inverse && !matched) || (!options.inverse && matched)) {
-          var result = line;
-          if (options.lineNumber) {
-            result = '' + (index + 1) + ':' + line;
+          var lineNumber = index + 1;
+          var result = {};
+          if (matches.length > 0) {
+            // If the last result intersects, combine them
+            var last = matches[matches.length - 1];
+            var minimumLineNumber = Math.max(
+              1,
+              lineNumber - options.beforeContext - 1
+            );
+            if (
+              last.hasOwnProperty('' + lineNumber) ||
+              last.hasOwnProperty('' + minimumLineNumber)
+            ) {
+              result = last;
+            }
+          }
+          result[lineNumber] = {
+            line: line,
+            match: true,
+          };
+          if (options.beforeContext > 0) {
+            // Store the lines with their line numbers to check for overlap
+            lines
+              .slice(Math.max(index - options.beforeContext, 0), index)
+              .forEach(function (v, i, a) {
+                var lineNum = '' + (index - a.length + i + 1);
+                if (!result.hasOwnProperty(lineNum)) {
+                  result[lineNum] = { line: v, match: false };
+                }
+              });
+          }
+          if (options.afterContext > 0) {
+            // Store the lines with their line numbers to check for overlap
+            lines
+              .slice(
+                index + 1,
+                Math.min(index + options.afterContext + 1, lines.length - 1)
+              )
+              .forEach(function (v, i) {
+                var lineNum = '' + (index + 1 + i + 1);
+                if (!result.hasOwnProperty(lineNum)) {
+                  result[lineNum] = { line: v, match: false };
+                }
+              });
+          }
+          // Only add the result if it's new
+          if (!matches.includes(result)) {
```

Comment:
This condition seems redundant to me. How could the result already be in the array if the object was created fresh at the start of the loop iteration (line 105)?

One way to double check this would be to temporarily change this to:

```
if (!matches.includes(result)) {
  matches.push(result);
} else {
  console.log('Do not submit this part. This line exists only to check if it shows up in the test coverage result');
}
```

Reply:
I forgot to mention, but you can run code coverage with `npm run test-with-coverage`. Be warned, this will run a bit slower than a regular `npm test`. You'll need to read through the coverage report carefully to see if this line number is in the "uncovered lines" column.

Reply:
On line 117, `result` is assigned to the last match, so it's no longer a new object. I've been using `test-with-coverage` while writing tests, and it confirms that this line is indeed being covered!